pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go

```diff
@@ -0,0 +1,497 @@
+		},
+	)
+	controller.claimStore, controller.claimController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return client.ListPersistentVolumeClaims(api.NamespaceAll, unversioned.ListOptions{})
+			},
+			WatchFunc: func(options unversioned.ListOptions) (watch.Interface, error) {
+				return client.WatchPersistentVolumeClaims(api.NamespaceAll, options)
+			},
+		},
+		&api.PersistentVolumeClaim{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    controller.handleAddClaim,
+			UpdateFunc: controller.handleUpdateClaim,
+			//			DeleteFunc: controller.handleDeleteClaim,
+		},
+	)
+
+	return controller, nil
+}
+
+func (controller *PersistentVolumeProvisionerController) handleAddVolume(obj interface{}) {
+	controller.locks["_main"].Lock()
+	defer controller.locks["_main"].Unlock()
+	cachedPv, _, _ := controller.volumeStore.Get(obj)
+	if pv, ok := cachedPv.(*api.PersistentVolume); ok {
+		err := controller.reconcileVolume(pv)
+		if err != nil {
+			glog.Errorf("Error encoutered reconciling volume %s: %+v", pv.Name, err)
+		}
+	}
+}
+
+func (controller *PersistentVolumeProvisionerController) handleUpdateVolume(oldObj, newObj interface{}) {
+	controller.handleAddVolume(newObj)
+}
+
+func (controller *PersistentVolumeProvisionerController) handleAddClaim(obj interface{}) {
+	controller.locks["_main"].Lock()
+	defer controller.locks["_main"].Unlock()
+	cachedPvc, exists, _ := controller.claimStore.Get(obj)
+	if !exists {
+		glog.Errorf("PersistentVolumeCache does not exist in the local cache: %+v", obj)
+		return
+	}
+	if pvc, ok := cachedPvc.(*api.PersistentVolumeClaim); ok {
+		err := controller.reconcileClaim(pvc)
+		if err != nil {
+			glog.Errorf("Error encoutered reconciling claim %s: %+v", pvc.Name, err)
+		}
+	}
+}
+
+func (controller *PersistentVolumeProvisionerController) handleUpdateClaim(oldObj, newObj interface{}) {
+	controller.handleAddClaim(newObj)
+}
+
+func (controller *PersistentVolumeProvisionerController) reconcileClaim(claim *api.PersistentVolumeClaim) error {
+	// no provisioning requested, return Pending. Claim may be pending indefinitely without a match.
+	if !keyExists(qosProvisioningKey, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] no provisioning required", claim.Name)
+		return nil
+	}
+	if len(claim.Spec.VolumeName) != 0 {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] already bound. No provisioning required", claim.Name)
+		return nil
+	}
+	if isAnnotationMatch(pvProvisioningRequiredAnnotationKey, pvProvisioningCompletedAnnotationValue, claim.Annotations) {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is already provisioned.", claim.Name)
+		return nil
+	}
+
+	// a quality-of-service annotation represents a specific kind of resource request by a user.
+	// the qos value is opaque to the system and will be configurable per plugin.
+	qos, _ := claim.Annotations[qosProvisioningKey]
+	if controller.provisioner != nil {
```

Comment:
suggest rewriting that as an early return if the provisioner is nil, then the rest of this code doesn't need to be indented.